### PR TITLE
fixed ethereals being unable to charge from APCs

### LINKED ISSUE #64485: Ethereals cannot charge from, or discharge into, APC's
## Reproduction: 
1. Be an Ethereal.
2. Attempt to charge yourself by right-clicking on a full APC on combat mode.
3. If you have no access to APC's: get Access Denied. If you do have access to APC's: unlocks/relocks the APC.
4. Charging doesn't happen.
Attempting to discharge yourself into an empty APC through right-clicking with combat mode off has the same effect.

This is almost certainly due to #64352, which moved lock/unlock APC from alt+click to right-click.

### LINKED ISSUE #64517: Ethereal , various problems
Reporting client version: 514.1575

## Round ID:
[177452](https://scrubby.melonmesa.com/round/177452)

## Testmerges:
- [Subsystems now track their avg and last tick allocations](https://www.github.com/tgstation/tgstation/pull/64448)

## Reproduction:

Mainly two problems since yesterday/today:
- I got blood Overcharge (charged too long in a cyborg charger) but the moodlet didn't actualize in time: I've gone to the sad face when you need to eat, at the smiling face when it was too late. Similar, clicking on the mental status should say negative things about the blood overcharge but I have the standard "I feel the power in my veins".
To clarify, I don't know if the Blood Overcharge hud icon wasn't supposed to happen, or if the mental status wasn't updated.

- Right Click on an APC should allow an Ethereal to "eat" electricity out of it. But since at least yesterday, it only lock/unlock/access denied. Non-fighting intent right click/left click does the same thing. Ethereal can't feed on APC anymore. I suppose that a bug.

EDIT: Second problem seems to be a byproduct of #64352 and mentioned in #64485 too.

### PR DESCRIPTION
**Pull request by MacBlaze1**
## About The Pull Request

This PR allows ethereals to receive and give charge to APCs again. I had to make some design choices due to the fact there is no clear solution with ethereals having an action with two possible outcomes. I moved ethereal APC charging to its own function called from the right click attack hand function.

## Why It's Good For The Game

APC charging is the main way ethereals gain nutrition and a previous PR made them unable to gain charge from nutrition.
This fixes that as best as possible by making them able to unlock and lock APCs above normal charge(No orange warning battery) with right click.

closes #64517
closes #64485

## Changelog
:cl:
fix: ethereals can now charge themselves using right click, but they will also lock/unlock the APC above normal charge
/:cl:

### CODE REVIEW

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -781,9 +781,64 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		var/obj/item/organ/stomach/maybe_stomach = H.getorganslot(ORGAN_SLOT_STOMACH)
+		if(istype(maybe_stomach, /obj/item/organ/stomach/ethereal))
+			var/obj/item/organ/stomach/ethereal/stomach = maybe_stomach
+			if(stomach.crystal_charge >= ETHEREAL_CHARGE_NORMAL)
+				togglelock(user)
+			ethereal_interact(user,modifiers)
+		else
+			togglelock(user)
```
Comment (tralezab) [inline code/modules/power/apc.dm:793]: shorter conditional first

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -781,9 +781,64 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
 	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
 
+/obj/machinery/power/apc/proc/ethereal_interact(mob/living/user,list/modifiers)
+	if(ishuman(user))
```
Comment (tralezab) [inline code/modules/power/apc.dm:797]: early return this

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -781,9 +781,64 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+		var/mob/living/carbon/human/H = user
+		var/obj/item/organ/stomach/maybe_stomach = H.getorganslot(ORGAN_SLOT_STOMACH)
+
+		if(istype(maybe_stomach, /obj/item/organ/stomach/ethereal))
```
Comment (tralezab) [inline code/modules/power/apc.dm:801]: early return this

File: code/modules/power/apc.dm [hunk 4]
```diff
@@ -781,9 +781,64 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+			var/charge_limit = ETHEREAL_CHARGE_DANGEROUS - APC_POWER_GAIN
+			var/obj/item/organ/stomach/ethereal/stomach = maybe_stomach
+			if((stomach?.drain_time < world.time) && LAZYACCESS(modifiers, RIGHT_CLICK))
```
Comment (tralezab) [inline code/modules/power/apc.dm:804]: early return this

File: code/modules/power/apc.dm [hunk 5]
```diff
@@ -781,9 +781,64 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
 		return
 	if(!user.canUseTopic(src, !issilicon(user)) || !isturf(loc))
 		return
-	togglelock(user)
+	if(ishuman(user))
```
Comment (tralezab) [inline code/modules/power/apc.dm:784]: early return this

**Review by tralezab — CHANGES_REQUESTED** [review]
if you don't know what that means, take a look at our style guide

File: code/modules/power/apc.dm (hunk 1 above)
Comment (MacBlaze1) [reply code/modules/power/apc.dm:793]: alright

File: code/modules/power/apc.dm (hunk 2 above)
Comment (MacBlaze1) [reply code/modules/power/apc.dm:797]: got it

File: code/modules/power/apc.dm (hunk 3 above)
Comment (MacBlaze1) [reply code/modules/power/apc.dm:801]: done

File: code/modules/power/apc.dm (hunk 4 above)
Comment (MacBlaze1) [reply code/modules/power/apc.dm:804]: got it

File: code/modules/power/apc.dm (hunk 5 above)
Comment (MacBlaze1) [reply code/modules/power/apc.dm:784]: Ill push now

File: code/modules/power/apc.dm [hunk 6]
```diff
@@ -781,9 +781,72 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+			var/obj/item/organ/stomach/ethereal/stomach = maybe_stomach
+			if(stomach.crystal_charge >= ETHEREAL_CHARGE_NORMAL)
+				togglelock(user)
+			ethereal_interact(user,modifiers)
 	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
 
+/obj/machinery/power/apc/proc/ethereal_interact(mob/living/user,list/modifiers)
+	if(!ishuman(user))
+		return
+	else
```
Comment (tralezab) [inline code/modules/power/apc.dm:801]: dude wtf- returning ends the proc, so why did you keep the else case. remove those and unindent every time you remove an else after the return

**Review by tralezab — CHANGES_REQUESTED** [review]
1-800-come-on-now

File: code/modules/power/apc.dm (hunk 6 above)
Comment (MacBlaze1) [reply code/modules/power/apc.dm:801]: its been a while okay, Its late, Im sorry

**Comment by RandomPredictable** [comment]
What about Ethereals being able to charge an empty APC with their own charge? Wouldn’t that feature (and the only way to lose charge if you’re overcharged, aside from gambling on the heart attack chance) get broken by the requirement to be below normal charge?

**Comment by MacBlaze1** [comment]
> What about Ethereals being able to charge an empty APC with their own charge? Wouldn’t that feature (and the only way to lose charge if you’re overcharged, aside from gambling on the heart attack chance) get broken by the requirement to be below normal charge?

no, it still works, I tested it myself.

File: code/modules/power/apc.dm [hunk 7]
```diff
@@ -781,9 +781,69 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
 		return
 	if(!user.canUseTopic(src, !issilicon(user)) || !isturf(loc))
 		return
-	togglelock(user)
+	if(!ishuman(user))
+		return
+	else
+		var/mob/living/carbon/human/H = user
```
Comment (tralezab) [inline code/modules/power/apc.dm:787]: bad var name, please make it descriptive

File: code/modules/power/apc.dm [hunk 8]
```diff
@@ -781,9 +781,69 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+			var/obj/item/organ/stomach/ethereal/stomach = maybe_stomach
+			if(stomach.crystal_charge >= ETHEREAL_CHARGE_NORMAL)
+				togglelock(user)
+			ethereal_interact(user,modifiers)
 	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
 
+/obj/machinery/power/apc/proc/ethereal_interact(mob/living/user,list/modifiers)
+	if(!ishuman(user))
+		return
+	var/mob/living/carbon/human/H = user
```
Comment (tralezab) [inline code/modules/power/apc.dm:801]: bad var name here

File: code/modules/power/apc.dm [hunk 9]
```diff
@@ -781,9 +781,69 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+	if(!((stomach?.drain_time < world.time) && LAZYACCESS(modifiers, RIGHT_CLICK)))
+		return
+	if(H.combat_mode)
+		if(cell.charge <= (cell.maxcharge / 2)) // ethereals can't drain APCs under half charge, this is so that they are forced to look to alternative power sources if the station is running low
+			to_chat(H, span_warning("The APC's syphon safeties prevent you from draining power!"))
+			return
+		if(stomach.crystal_charge > charge_limit)
+			to_chat(H, span_warning("Your charge is full!"))
+			return
+		stomach.drain_time = world.time + APC_DRAIN_TIME
```
Comment (tralezab) [inline code/modules/power/apc.dm:817]: this is done on both modes, do it after each case. so if() > code, else > code, after both put in `stomach.drain_time = world.time + APC_DRAIN_TIME`

File: code/modules/power/apc.dm [hunk 10]
```diff
@@ -781,9 +781,69 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/power/apc/auto_name, APC_PIXEL_OFFSET
...
+		to_chat(H, span_notice("You start channeling some power through the APC into your body."))
+		if(do_after(user, APC_DRAIN_TIME, target = src))
+			if(cell.charge <= (cell.maxcharge / 2) || (stomach.crystal_charge > charge_limit))
+				return
+			to_chat(H, span_notice("You receive some charge from the APC."))
+			stomach.adjust_charge(APC_POWER_GAIN)
+			cell.charge -= APC_POWER_GAIN
+		return
```
Comment (tralezab) [inline code/modules/power/apc.dm:825]: trailing return, the proc ends after this anyways

**Review by tralezab — COMMENTED** [review]
i messed up, when you go to make a fix refer to this

**Pull request merged**